src/auth/src/credentials/impersonated.rs

```diff
@@ -550,6 +594,51 @@ pub(crate) fn build_components_from_json(
     })
 }
 
+#[cfg(google_cloud_unstable_signed_url)]
+// Check if the source credentials is a service account and if so, build a signer from it.
+// Service account signers can do signing locally, which is more efficient than using the
+// remote/API based signer.
+fn build_signer_from_json(json: Value) -> BuildResult<Option<crate::signer::Signer>> {
+    use crate::credentials::service_account::ServiceAccountKey;
+    use crate::signer::service_account::ServiceAccountSigner;
+
+    let config =
+        serde_json::from_value::<ImpersonatedConfig>(json).map_err(BuilderError::parsing)?;
+
+    let client_email = extract_client_email(&config.service_account_impersonation_url)?;
+    let source_credential_type = extract_credential_type(&config.source_credentials)?;
+    if source_credential_type == "service_account" {
+        let service_account_key =
+            serde_json::from_value::<ServiceAccountKey>(config.source_credentials)
+                .map_err(BuilderError::parsing)?;
+        let signing_provider = ServiceAccountSigner::from_impersonated_service_account(
+            service_account_key,
+            client_email,
+        );
+        let signer = crate::signer::Signer {
+            inner: Arc::new(signing_provider),
+        };
+        return Ok(Some(signer));
+    }
+    Ok(None)
+}
+
+#[cfg(google_cloud_unstable_signed_url)]
+fn extract_client_email(service_account_impersonation_url: &str) -> BuildResult<String> {
+    let parts: Vec<&str> = service_account_impersonation_url
+        .split("serviceAccounts/")
```

Comment:
Shouldn't this be `/serviceAccounts/` with a leading slash?

Reply:
good catch